### PR TITLE
Add Clippy

### PR DESCRIPTION
**Pull request by sunjay**
[Clippy](https://github.com/rust-lang/rust-clippy) is a fun tool designed to help us improve the quality of our code. If you've used a "linter" in other languages, clippy is Rust's "linter". Unlike rustfmt which checks/enforces a certain *style* in our codebase, clippy's job is to help us avoid common mistakes. 

It does this by printing helpful messages about potential improvements we can make:

```
warning: this .into_iter() call is equivalent to .iter() and will not move the slice
  --> src/main.rs:85:30
   |
85 |     for (i, frame) in frames.into_iter().enumerate() {
   |                              ^^^^^^^^^ help: call directly: `iter`
   |
   = note: #[warn(clippy::into_iter_on_ref)] on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#into_iter_on_ref

warning: this .into_iter() call is equivalent to .iter() and will not move the slice
   --> src/main.rs:119:30
    |
119 |     for (i, frame) in frames.into_iter().enumerate() {
    |                              ^^^^^^^^^ help: call directly: `iter`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#into_iter_on_ref
```

These messages are often very helpful and detailed. You can learn a lot about writing idiomatic Rust from using clippy.

That being said, and as you know from our testing course, all static analysis tools are inherently limited. Clippy is no exception. It can very opinionated at times and this can lead to it warning about things that aren't even really issues. The tool policy (see below) is designed to address that.

## Running clippy locally

To run clippy, you will need to "add" it to your rust installation:

```bash
rustup component add clippy # only need to do this once to install clippy
```

Then you can run clippy using its cargo command:

```bash
cargo clippy
```

## Tool Policy

To deal with the inaccuracies of these tools, I have proposed and documented a "tool policy" to help remind us about why we have these tools and to give *everyone* the freedom to configure them.

The policy is documented at the top of `main.rs`, but I've reproduced it here for your convenience:

- We add tools in order to help *us* improve our code
- If they are not doing that, we can configure them or even elect to remove them
- No tool is perfect and we are allowed to disagree with its results
- If the tool warns about something that isn't actually an issue worth caring about, add it to
  the list below and explain your change in your PR
- We don't want to litter our code with #[allow] attributes unnecessarily, so try to either
  globally disable that aspect of the tool or live with it and do what the tool says
- If we make a mistake and find that one of these lints shouldn't have been added here, we can
  always remove it later

See `main.rs` to understand what "below" means.

---

This policy is carefully worded to empower *everyone* in our group to use your own judgment and not apply the suggestions from clippy blindly.

cc @ProtoArt/core-team

### CODE REVIEW

**Comment by sunjay** [comment]
@ProtoArt/core-team please read this PR and then "approve" it so we know  you've read it. Once everyone has approved, we'll merge it in.

**Review by JLSJamesShi — APPROVED** [review]
very good, much wow

**Review by daose — APPROVED** [review]

**Comment by sunjay** [comment]
Rebased from latest master and applied remaining clippy fixes. @marauder00 please read & approve when you have a chance. :)

**Comment by sunjay** [comment]
The `-D warnings` we had in our Travis CI config will deny all warnings, even the ones from rustc. We really only want clippy warnings to be denied though, so my last commit updates things to fix that.

(I also submitted a PR to clippy's repo so that the docs cover this as well)

**Review by marauder00 — APPROVED** [review]
THANK YOU FOR EMPOWERING EVERY PERSON IN THIS GitHub ORGANIZATION ON THIS PLANET TO ACHIEVE MOREEE!!!

**Pull request merged**
